src/state_awset.erl

```diff
@@ -0,0 +1,468 @@
+        Error         -> Error
+    end.
+
+%% ===================================================================
+%% EUnit tests
+%% ===================================================================
+-ifdef(TEST).
+
+new_test() ->
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, orddict:new()}, ordsets:new()}},
+                 new()).
+
+query_test() ->
+    Set0 = new(),
+    Set1 = {?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{a, 2}]}}]},
+                    [{a, 1}, {a, 2}]}},
+    ?assertEqual(sets:new(), query(Set0)),
+    ?assertEqual(sets:from_list([<<"a">>]), query(Set1)).
+
+delta_add_test() ->
+    Actor = 1,
+    Set0 = new(),
+    {ok, {?TYPE, {delta, Delta1}}} = delta_mutate({add, <<"a">>}, Actor, Set0),
+    Set1 = merge({?TYPE, Delta1}, Set0),
+    {ok, {?TYPE, {delta, Delta2}}} = delta_mutate({add, <<"a">>}, Actor, Set1),
+    Set2 = merge({?TYPE, Delta2}, Set1),
+    {ok, {?TYPE, {delta, Delta3}}} = delta_mutate({add, <<"b">>}, Actor, Set2),
+    Set3 = merge({?TYPE, Delta3}, Set2),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{1, 1}]}}]},
+                          [{1, 1}]}},
+                 {?TYPE, Delta1}),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{1, 1}]}}]},
+                          [{1, 1}]}},
+                 Set1),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}}]},
+                          [{1, 1}, {1, 2}]}},
+                 {?TYPE, Delta2}),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}}]},
+                          [{1, 1}, {1, 2}]}},
+                 Set2),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, [{<<"b">>, {dot_set, [{1, 3}]}}]},
+                          [{1, 3}]}},
+                 {?TYPE, Delta3}),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}},
+                            {<<"b">>, {dot_set, [{1, 3}]}}]},
+                          [{1, 1}, {1, 2}, {1, 3}]}},
+                 Set3).
+
+add_test() ->
+    Actor = 1,
+    Set0 = new(),
+    {ok, Set1} = mutate({add, <<"a">>}, Actor, Set0),
+    {ok, Set2} = mutate({add, <<"a">>}, Actor, Set1),
+    {ok, Set3} = mutate({add, <<"b">>}, Actor, Set2),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{1, 1}]}}]},
+                          [{1, 1}]}},
+                 Set1),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}}]},
+                          [{1, 1}, {1, 2}]}},
+                 Set2),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}},
+                            {<<"b">>, {dot_set, [{1, 3}]}}]},
+                          [{1, 1}, {1, 2}, {1, 3}]}},
+                 Set3).
+
```

Comment:
Same fixes could apply here. Notice that this is not incorrect as it is, but it uses much more meta-data than needed in the long run.